packages/event-listener/src/index.ts

```diff
@@ -1,128 +1,135 @@
-import { Accessor, createEffect, onCleanup } from "solid-js";
-import type { JSX } from "solid-js";
+import { accessAsArray, Fn, isClient, Many, MaybeAccessor } from "@solid-primitives/utils";
+import { Accessor, createEffect, JSX, onCleanup } from "solid-js";
+import { createCallbackStack } from "./common";
+export * from "./component";
 
-export type EventMapOf<T> = T extends Window
+export type EventMapOf<Target> = Target extends Window
   ? WindowEventMap
-  : T extends Document
+  : Target extends Document
   ? DocumentEventMap
-  : T extends HTMLElement
+  : Target extends HTMLElement
   ? HTMLElementEventMap
-  : T extends MediaQueryList
+  : Target extends MediaQueryList
   ? MediaQueryListEventMap
-  : {};
-
-export type EventMapOfMultiple<T> = T extends EventTarget
-  ? EventMapOf<T>
-  : T extends EventTarget[]
-  ? EventMapOf<T[number]>
   : never;
 
-export type EventListenerProps<
-  T extends EventTarget | EventTarget[],
-  E extends Record<string, Event> = {}
-> = [
-  name: [{}, {}] extends [EventMapOfMultiple<T>, E]
-    ? string
-    : string & (keyof EventMapOfMultiple<T> | keyof E),
-  handler: EventListenerOrEventListenerObject | null,
-  options?: AddEventListenerOptions
-];
+export type EventHandler<
+  EventMap extends
+    | HTMLElementEventMap
+    | WindowEventMap
+    | DocumentEventMap
+    | MediaQueryListEventMap
+    | Record<string, Event>,
+  EventName extends keyof EventMap
+> = (event: EventMap[EventName]) => void;
+
+export type EventListenerReturn = [stop: Fn, start: Fn];
 
-export type CreateEventListenerReturn = [
-  add: (el: EventTarget) => void,
-  remove: (el: EventTarget) => void
-];
+export type EventListenerDirectiveProps =
+  | [string, (e: any) => void, boolean | AddEventListenerOptions]
```

Comment:
you can use a named tuple here

```ts
[name: string, handler: (e: any) => void, options?: AddEventListenerOptions]
```

This is simpler to understand and also allows the use of optionals.